### PR TITLE
Add test fixtures to the `compiler-utils` module to share utilities helpful for testing

### PR DESCRIPTION
**Pull request by vRallev**

### CODE REVIEW

File: compiler-utils/README.md [hunk 1]
```diff
@@ -2,4 +2,10 @@
...
 Contains several utility functions used within Anvil and external `CodeGenerator`s. There's no 
 API stability guaranteed for these functions. They can be changed or removed in every single
-release. So use at your own risk. 
+release. They're not documented, and the behavior can change. So use at your own risk. 
+
+The artifact comes with testing utilities for easier end-to-end tests for `CodeGenerators`. You 
+can add the testing utilities to your project through test fixtures:
+```groovy
+testImplementation testFixtures('com.squareup.anvil:compiler-utils:VERSION')
+```
```
Comment (vRallev) [inline compiler-utils/README.md:11]: A little bit of documentation.

File: compiler-utils/build.gradle [hunk 2]
```diff
@@ -1,10 +1,13 @@
 apply plugin: 'org.jetbrains.kotlin.jvm'
+apply plugin: 'java-test-fixtures'
```
Comment (vRallev) [inline compiler-utils/build.gradle:2]: I'm using test fixtures for testing utilities. That seemed to be a good use case here.

File: compiler-utils/build.gradle [hunk 3]
```diff
@@ -13,4 +16,29 @@ dependencies {
...
+  testFixturesImplementation deps.junit
+  testFixturesImplementation deps.truth
+
+  // This workaround is needed to resolve classes in the IDE properly.
+  if (isIdeSync) {
+    compileOnly deps.dagger2.compiler
+    compileOnly deps.junit
+    compileOnly deps.kotlin.compile_testing
+    compileOnly deps.truth
+  }
```
Comment (vRallev) [inline compiler-utils/build.gradle:34]: .. If only the IDE would like them 😏

File: integration-tests/code-generator/src/test/java/com/squareup/anvil/test/TestCodeGeneratorTest.kt [hunk 4]
```diff
@@ -0,0 +1,29 @@
+package com.squareup.anvil.test
+
+import com.google.common.truth.Truth.assertThat
+import com.sqareup.anvil.compiler.internal.testing.compile
+import com.sqareup.anvil.compiler.internal.testing.createInstance
+import org.junit.Test
+
+class TestCodeGeneratorTest {
```
Comment (vRallev) [inline integration-tests/code-generator/src/test/java/com/squareup/anvil/test/TestCodeGeneratorTest.kt:8]: I've added this integration test to make sure it works as expected.

File: integration-tests/tests/build.gradle [hunk 5]
```diff
@@ -8,6 +8,7 @@ kotlin {
 
 dependencies {
   testImplementation project(':integration-tests:library')
+  testImplementation testFixtures(project(":compiler-utils"))
```
Comment (vRallev) [inline integration-tests/tests/build.gradle:11]: I'm reusing the testing utilities in several modules now.

**Comment by vRallev** [comment]
Hm, seeing this in the log:
```
Maven publication 'maven' pom metadata warnings (silence with 'suppressPomMetadataWarningsFor(variant)'):
  - Variant testFixturesApiElements:
      - Declares capability com.squareup.anvil:compiler-utils-test-fixtures:2.2.4-SNAPSHOT which cannot be mapped to Maven
  - Variant testFixturesRuntimeElements:
      - Declares capability com.squareup.anvil:compiler-utils-test-fixtures:2.2.4-SNAPSHOT which cannot be mapped to Maven
These issues indicate information that is lost in the published 'pom' metadata file, which may be an issue if the published library is consumed by an old Gradle version or Apache Maven.
The 'module' metadata file, which is used by Gradle 6+ is not affected.
```
Not sure if I should be worried. Probably not due to the last sentence. I don't care about Gradle 5 or older.

File: compiler-utils/README.md (hunk 1 above)
Comment (0xMatthewGroves) [reply compiler-utils/README.md:11]: Neat. I had no idea testFixtures existed.

File: compiler-utils/src/testFixtures/java/com/sqareup/anvil/compiler/internal/testing/TestUtils.kt [hunk 6]
```diff
@@ -0,0 +1,181 @@
...
+  allWarningsAsErrors: Boolean = true,
+  useIR: Boolean = true,
+  messageOutputStream: OutputStream = System.out,
+  block: Result.() -> Unit = { }
+): Result {
+  return KotlinCompilation()
+    .apply {
+      compilerPlugins = listOf(AnvilComponentRegistrar())
+      this.useIR = useIR
+      useOldBackend = !useIR
```
Comment (0xMatthewGroves) [inline compiler-utils/src/testFixtures/java/com/sqareup/anvil/compiler/internal/testing/TestUtils.kt:42]: Is there a backward compatibility reason for keeping this around? No way of just using the new IR by default?

File: compiler-utils/src/testFixtures/java/com/sqareup/anvil/compiler/internal/testing/TestUtils.kt [hunk 7]
```diff
@@ -0,0 +1,181 @@
...
+        val packageDir = content.lines()
+          .firstOrNull { it.trim().startsWith("package ") }
+          ?.substringAfter("package ")
+          ?.replace('.', '/')
+          ?.let { "$it/" }
+          ?: ""
+
+        val name = "${workingDir.absolutePath}/sources/src/main/java/$packageDir/Source.kt"
+        with(File(name).parentFile) {
+          check(exists() || mkdirs())
```
Comment (0xMatthewGroves) [inline compiler-utils/src/testFixtures/java/com/sqareup/anvil/compiler/internal/testing/TestUtils.kt:89]: Shorthand for make directory if not exists?

File: compiler-utils/src/testFixtures/java/com/sqareup/anvil/compiler/internal/testing/TestUtils.kt [hunk 8]
```diff
@@ -0,0 +1,181 @@
...
+public fun Class<*>.moduleFactoryClass(
+  providerMethodName: String,
+  companion: Boolean = false
+): Class<*> {
+  val companionString = if (companion) "_Companion" else ""
+  val enclosingClassString = enclosingClass?.let { "${it.simpleName}_" } ?: ""
+
+  return classLoader.loadClass(
+    "${packageName()}$enclosingClassString$simpleName$companionString" +
+      "_${providerMethodName.capitalize(Locale.US)}Factory"
```
Comment (0xMatthewGroves) [inline compiler-utils/src/testFixtures/java/com/sqareup/anvil/compiler/internal/testing/TestUtils.kt:109]: Didn't you have an extension override for this one?

File: compiler-utils/src/testFixtures/java/com/sqareup/anvil/compiler/internal/testing/TestUtils.kt [hunk 9]
```diff
@@ -0,0 +1,181 @@
...
+    .also { assertThat(it).hasSize(1) }
+    .first()
+
+@ExperimentalAnvilApi
+public val Class<*>.daggerModule: Module
+  get() = annotations.filterIsInstance<Module>()
+    .also { assertThat(it).hasSize(1) }
+    .first()
+
+@ExperimentalAnvilApi
```
Comment (0xMatthewGroves) [inline compiler-utils/src/testFixtures/java/com/sqareup/anvil/compiler/internal/testing/TestUtils.kt:158]: Love the extensions usage in this file 👍

**Review by 0xMatthewGroves — APPROVED** [review]
LGTM. I appreciated the PR comments as well.

File: compiler-utils/src/testFixtures/java/com/sqareup/anvil/compiler/internal/testing/AnyDaggerComponent.kt [hunk 10]
```diff
@@ -1,15 +1,18 @@
-package com.squareup.anvil.compiler
+package com.sqareup.anvil.compiler.internal.testing
```
Comment (lmchew) [inline compiler-utils/src/testFixtures/java/com/sqareup/anvil/compiler/internal/testing/AnyDaggerComponent.kt:1]: Oops, you went from squareup --> sqareup here.

File: compiler-utils/src/testFixtures/java/com/sqareup/anvil/compiler/internal/testing/ReflectionTestUtils.kt [hunk 11]
```diff
@@ -1,22 +1,29 @@
-package com.squareup.anvil.compiler
+@file:Suppress("unused")
 
+package com.sqareup.anvil.compiler.internal.testing
```
Comment (lmchew) [inline compiler-utils/src/testFixtures/java/com/sqareup/anvil/compiler/internal/testing/ReflectionTestUtils.kt:3]: Another sqareup here.

File: compiler-utils/src/testFixtures/java/com/sqareup/anvil/compiler/internal/testing/TestUtils.kt [hunk 12]
```diff
@@ -0,0 +1,181 @@
+@file:Suppress("unused")
+
+package com.sqareup.anvil.compiler.internal.testing
```
Comment (lmchew) [inline compiler-utils/src/testFixtures/java/com/sqareup/anvil/compiler/internal/testing/TestUtils.kt:3]: Another sqareup.

**Review by lmchew — APPROVED** [review]

File: compiler-utils/src/testFixtures/java/com/sqareup/anvil/compiler/internal/testing/TestUtils.kt (hunk 6 above)
Comment (vRallev) [reply compiler-utils/src/testFixtures/java/com/sqareup/anvil/compiler/internal/testing/TestUtils.kt:42]: We're still using the old backend in our codebase I'm sure others do as well. I want to keep compatibility for a little longer until I rip out the code.

File: compiler-utils/src/testFixtures/java/com/sqareup/anvil/compiler/internal/testing/TestUtils.kt (hunk 7 above)
Comment (vRallev) [reply compiler-utils/src/testFixtures/java/com/sqareup/anvil/compiler/internal/testing/TestUtils.kt:89]: Yes, pretty much :)

File: compiler-utils/src/testFixtures/java/com/sqareup/anvil/compiler/internal/testing/TestUtils.kt (hunk 8 above)
Comment (vRallev) [reply compiler-utils/src/testFixtures/java/com/sqareup/anvil/compiler/internal/testing/TestUtils.kt:109]: Yes, that's the Kotlin 1.5 PR that I still did not merge. But I rebased it already on top of this one (which was easier to resolve for merge conflicts) and I'll address there.

File: compiler-utils/src/testFixtures/java/com/sqareup/anvil/compiler/internal/testing/TestUtils.kt (hunk 12 above)
Comment (vRallev) [reply compiler-utils/src/testFixtures/java/com/sqareup/anvil/compiler/internal/testing/TestUtils.kt:3]: Oh damn, nice catch. I'll address this in a follow up PR. Stacking is annoying sometimes.

Comment (vRallev) [reply compiler-utils/src/testFixtures/java/com/sqareup/anvil/compiler/internal/testing/TestUtils.kt:3]: https://github.com/square/anvil/pull/293

**Pull request merged**